### PR TITLE
Add references and byte support in bytecode

### PR DESCRIPTION
**Pull request by Override-6**
This pull request replaces the current `SET_LOCAL` and `GET_LOCAL` by 3 `GET/SET` opcodes for byte, quad-words and references.

The constant pool has been refactored, and now all the strings (runtimes and constants) are owned by a set which lives during all the vm execution.

Also added `VirtualMachineError`, with two subdivisions `InvalidBytecodeError` and `MemoryError`.

### CODE REVIEW

File: vm/src/memory/constant_pool.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+template <typename T>
+class PoolConstantWrappedValue : public PoolConstantValue {
+public:
+    const T value;
+    explicit PoolConstantWrappedValue(PoolConstantType type, T value) : PoolConstantValue{type}, value{std::move(value)} {}
+};
+
+std::unique_ptr<PoolConstantWrappedValue<const std::string *>> read_string(ByteReader &reader, strings_t &strings) {
+    // Read the length
+    size_t length = ntohl(reader.read<size_t>());
```
Comment (syldium) [inline vm/src/memory/constant_pool.cpp:18]: `size_t` is architecture dependent.

File: vm/src/memory/constant_pool.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    // Allocate the string
+    std::string str(reader.read_n<char>(length), length);
+
+    auto [str_it, _] = strings.insert(std::make_unique<std::string>(str));
```
Comment (syldium) [inline vm/src/memory/constant_pool.cpp:23]:
```suggestion
    std::string str(reader.read_n<char>(length), length);

    auto [str_it, _] = strings.insert(std::make_unique<std::string>(std::move(str)));
```

You are creating two strings, one by reading, and the other by calling `std::make_unique()`, that forward the arguments to the constructor. It is here the copy constructor.

File: vm/src/memory/constant_pool.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+class PoolConstantValue {
+public:
+    PoolConstantType type;
+    explicit PoolConstantValue(PoolConstantType type);
+};
+
+class BadConstantType : public VirtualMachineError {
+public:
+    explicit BadConstantType(std::string msg): VirtualMachineError(msg) {}
+};
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:33]: I'm skeptical about what other types could contain this pool.

**Review by syldium — CHANGES_REQUESTED** [review]

File: vm/src/memory/constant_pool.h (hunk 3 above)
Comment (Override-6) [reply vm/src/memory/constant_pool.h:33]: signatures

File: vm/src/errors.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <exception>
+#include <string>
+
+/**
+ * Base class for fatal errors that occurs inside the virtual machine
+ * */
+class VirtualMachineError: public std::exception {
```
[the remark this answers is not in the document]
Comment (Override-6) [reply vm/src/errors.h:9]: loading a constant pool isn't a runtime error, there is no real runtime errors currently, the only error thrown during interpreter is if the bytecode is invalid (wrong opcode), this is not really a _runtime_ error

File: compiler/src/type.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+use analyzer::types::hir::TypeId;
+use analyzer::types::*;
+
+/// returns the size of a given type identifier
+pub fn get_type_size(tpe: TypeId) -> TypeSize {
```
[the remark this answers is not in the document]
Comment (Override-6) [reply compiler/src/type.rs:5]: we just need to differenciate the primitives, the use of a higher type isn't worth it

File: compiler/src/bytecode.rs [hunk 6]
```diff
@@ -23,14 +25,30 @@ impl Bytecode {
...
+    /// assuming the local's size is the given `size` argument
+    pub fn emit_set_local(&mut self, identifier: LocalId, size: ValueStackSize) {
+        let opcode = match size {
+            ValueStackSize::Byte => Opcode::SetByte,
+            ValueStackSize::QWord => Opcode::SetQWord,
+            ValueStackSize::Reference => Opcode::SetRef,
+            ValueStackSize::Zero => panic!("set_local for value whose type is zero-sized"),
+        };
+        self.emit_code(opcode);
+        self.bytes.push(identifier.0 as u8);
```
Comment (syldium) [inline compiler/src/bytecode.rs:38]:
```suggestion
        self.bytes
            .push(u8::try_from(identifier.0).expect("local identifier too large"));
```

File: compiler/src/bytecode.rs [hunk 7]
```diff
@@ -23,14 +25,30 @@ impl Bytecode {
...
+    /// assuming the local's size is the given `size` argument
+    pub fn emit_get_local(&mut self, identifier: LocalId, size: ValueStackSize) {
+        let opcode = match size {
+            ValueStackSize::Byte => Opcode::GetByte,
+            ValueStackSize::QWord => Opcode::GetQWord,
+            ValueStackSize::Reference => Opcode::GetRef,
+            ValueStackSize::Zero => panic!("get_local for value whose type is zero-sized"),
+        };
+        self.emit_code(opcode);
+        self.bytes.push(identifier.0 as u8);
```
Comment (syldium) [inline compiler/src/bytecode.rs:51]:
```suggestion
        self.bytes
            .push(u8::try_from(identifier.0).expect("local identifier too large"));
```

File: vm/src/byte_reader.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * */
+class ByteReader {
+    const char *bytes;
+    const size_t byte_count;
+    size_t pos;
+
+public:
+    ByteReader(const char *bytes, size_t byte_count);
+
+    ByteReader(const ByteReader &other) = delete;
```
Comment (syldium) [inline vm/src/byte_reader.h:17]:
```suggestion
    ByteReader(const ByteReader &other) = delete;

    ByteReader &operator=(const ByteReader &other) = delete;
```

File: vm/src/memory/constant_pool.h [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+ * defines the type for an index inside the constant pool.
+ * */
+typedef uint32_t constant_index;
+
+/**
+ * Kind of constant that can be contained in the constant pool
+ * */
+enum PoolConstantType {
+    /// an utf-8 string
+    C_STR,
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:21]:
```suggestion
    UTF8_STR,
```
The string isn't encoded as a C string.

File: vm/src/memory/constant_pool.h [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+
+/**
+ * defines the type for an index inside the constant pool.
+ * */
+typedef uint32_t constant_index;
+
+/**
+ * Kind of constant that can be contained in the constant pool
+ * */
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:18]:
```suggestion
 */
```

File: vm/src/memory/constant_pool.h [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+#include "memory/strings.h"
+#include <memory>
+#include <vector>
+
+class ByteReader;
+
+
+/**
+ * defines the type for an index inside the constant pool.
+ * */
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:13]:
```suggestion
 */
```

File: vm/src/memory/constant_pool.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+class PoolConstantValue {
+public:
+    PoolConstantType type;
+    explicit PoolConstantValue(PoolConstantType type);
+};
+
+class BadConstantType : public VirtualMachineError {
+public:
+    explicit BadConstantType(std::string msg): VirtualMachineError(msg) {}
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:32]:
```suggestion
    explicit BadConstantType(std::string msg): VirtualMachineError(std::move(msg)) {}
```

File: vm/src/memory/constant_pool.cpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        throw InvalidBytecodeError("Error when reading module bytecode: " + std::string(e.what()));
+    }
+}
+
+template <typename T>
+const T &ConstantPool::get(constant_index pos, PoolConstantType type, const char *type_name) const {
+    const PoolConstantValue *value = constants.at(pos).get();
+    if (value->type != type) {
+        throw BadConstantType("attempted to access " + std::string(type_name) + " in constant pool at index " + std::to_string(pos) + " which is not a string");
+    }
```
Comment (syldium) [inline vm/src/memory/constant_pool.cpp:50]:
```suggestion
const T &ConstantPool::get(constant_index pos) const {
    static_assert(std::is_same_v<T, const std::string *>, "Type cannot be used in constant pool");
    constexpr PoolConstantType type = []() {
        if constexpr (std::is_same_v<T, const std::string *>) {
            return PoolConstantType::C_STR;
        }
    }();

    const PoolConstantValue *value = constants.at(pos).get();
    if (value->type != type) {
        std::stringstream ss;
        ss << "attempted to access " << type << " in constant pool at index " << pos << " which is not a string";
        throw BadConstantType(ss.str());
    }
```
`#include <sstream>`

File: vm/src/memory/constant_pool.h [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * defines the type for an index inside the constant pool.
+ * */
+typedef uint32_t constant_index;
+
+/**
+ * Kind of constant that can be contained in the constant pool
+ * */
+enum PoolConstantType {
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:19]:
```suggestion
enum class PoolConstantType {
```

File: vm/src/memory/constant_pool.h [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    /// an utf-8 string
+    C_STR,
+};
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:22]:
```suggestion
};

std::ostream &operator<<(std::ostream &os, const PoolConstantType &type);
```

Sample implementation:
```cpp
std::ostream &operator<<(std::ostream &os, const PoolConstantType &type) {
    switch (type) {
    case PoolConstantType::C_STR:
        return os << "C_STR";
    default:
        return os << "UNKNOWN";
    }
}
```

**Review by syldium — CHANGES_REQUESTED** [review]
Please format the code.

File: vm/src/memory/constant_pool.h [hunk 16]
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "errors.h"
+#include "memory/strings.h"
+#include <memory>
+#include <vector>
```
Comment (syldium) [inline vm/src/memory/constant_pool.h:6]:
```suggestion
#include <memory>
#include <vector>

#include "memory/strings.h"
```

File: vm/src/memory/constant_pool.cpp [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+            pool.constants[i] = value;
+        }
+        return pool;
+    } catch (std::out_of_range e) {
+        throw InvalidBytecodeError("Error when reading module bytecode: " + std::string(e.what()));
+    }
+}
+
+ConstantPool::ConstantPool(uint32_t size)
+    : constants{static_cast<std::string const **>(malloc(size * sizeof(std::string *)))},
```
Comment (syldium) [inline vm/src/memory/constant_pool.cpp:36]:
```suggestion
    : constants{std::make_unique<const std::string*[]>(size)},
```
The memory is never freed.

File: vm/CMakeLists.txt [hunk 18]
```diff
@@ -11,8 +11,10 @@ if (MSVC)
     set(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} /W4 /WX")
 endif()
 
-add_library(vm src/interpreter.cpp src/vm.cpp src/conversions.cpp src/memory/operand_stack.cpp src/pool.cpp)
-add_executable(vm_exe src/main.cpp)
+include_directories(src)
+
+add_library(vm src/interpreter.cpp src/vm.cpp src/memory/operand_stack.cpp src/memory/constant_pool.cpp src/byte_reader.cpp src/errors.cpp)
+add_executable(vm_exe src/main.cpp src/memory/constant_pool.cpp src/byte_reader.cpp src/errors.cpp)
```
Comment (syldium) [inline vm/CMakeLists.txt:17]:
```suggestion
add_executable(vm_exe src/main.cpp)
```

File: vm/src/vm.cpp [hunk 19]
```diff
@@ -1,38 +1,18 @@
...
+    strings_t strings;
 
     try {
-        run(std::move(pool), bytes + constant_pool_bytes, byte_count - constant_pool_bytes);
-    } catch (const std::exception &e) {
+        ByteReader reader(bytes, byte_count);
+        ConstantPool pool = load_constant_pool(reader, strings);
+
+        run(pool, bytes + reader.position(), byte_count - reader.position(), strings);
+    } catch (std::exception e) {
```
Comment (syldium) [inline vm/src/vm.cpp:15]:
```suggestion
    } catch (const std::exception &e) {
```

**Review by syldium — CHANGES_REQUESTED** [review]

**Review by syldium — APPROVED** [review]

**Pull request merged**